### PR TITLE
Don't keep unnecessary saved_inputs alive

### LINKED ISSUE #16577: Dropout2d uses a lot of memory.
## 🐛 Bug
`Dropout2d` uses a lot of memory. For a `B, F, H, W` tensor, it could get away with `B * F` bits to store whether each feature/batch item was kept or not. Instead, it appears to store something proportional to `B * F * H * W`. According to @apaszke this is because it is implemented with multiplication rather than `masked_fill` or `where` (since this makes it easier to switch between `eval` and `train`). A solution would be to have different control paths for `train` and `eval` or
> an easy fix would be to improve the derivative code we generate for multiplies such that they don’t save `x` when you do `x * y` with `y` that doesn’t need gradient

## Environment

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 396.54
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] _tflow_select             2.3.0                       mkl  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl_fft                   1.0.6            py36hd81dba3_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] pytorch                   1.0.0           py3.6_cuda9.0.176_cudnn7.4.1_1    pytorch
[conda] tensorflow                1.12.0          mkl_py36h69b6ba0_0  
[conda] tensorflow-base           1.12.0          mkl_py36h3c3e929_0  
[conda] torchvision               0.2.1                      py_2    pytorch

### PR DESCRIPTION
**Pull request by apaszke**
Fixes #16577.

This greatly improves memory efficiency of certain ops like Dropout2d. Previously, they were implemented as `input * mask` where mask never requires_grad, but we didn't use that knowledge in forward, and (in case of a in-place dropout) kept input.clone() for the backward, when it would simply get ignored.

This patch tries to address this situation by emitting some guards for stores like this, but only if they are as simple, as checking if a single value requires_grad.

Interestingly, the same optimizations apply to methods like bmm, baddmm, etc., but _not to mm nor addmm_, because of how their derivatives are defined. Apparently they unnecessarily use `mat1` to compute the derivative of `mat1` just to improve the error message in case `mat1` was sparse. I'd like to apply this optimization to that case, but I don't want to loose the nicer error message, so if anyone has any ideas for solutions, please let me know...

Full list of operators affected by this patch:
* _nnpack_spatial_convolution
* addbmm
* addcdiv
* addcmul
* addmv
* addr
* baddbmm
* bmm
* cross
* div
* dot
* fmod
* ger
* index_add_
* mul
* mv
* scatter_add_

### CODE REVIEW

**Comment by SsnL** [comment]
> I'd like to apply this optimization to that case, but I don't want to loose the nicer error message, so if anyone has any ideas for solutions, please let me know...

Idea: save `TensorGeometry(mat1)` instead, and augment `TensorGeometry` to include sparsity if needed.

**Comment by SsnL** [comment]
This looks awesome. A before & after comparison on generated code would be great! :)

**Comment by apaszke** [comment]
Oh yes, I forgot to show the example change. Here's the case of `mul_` (the one we use in in-place dropout).

Old code:
```cpp
ensor & VariableType::mul_(Tensor & self, const Tensor & other) const {
  ...
  if (compute_requires_grad( self, other )) {
    grad_fn = std::shared_ptr<MulBackward0>(new MulBackward0(), deleteFunction);
    grad_fn->set_next_edges(collect_next_edges( self, other ));
    grad_fn->self_ = SavedVariable(self.clone(), false);
    grad_fn->other_ = SavedVariable(other, false);
  }
  ...
}
```

New code:
```cpp
Tensor & VariableType::mul_(Tensor & self, const Tensor & other) const {
  ...
  if (compute_requires_grad( self, other )) {
    grad_fn = std::shared_ptr<MulBackward0>(new MulBackward0(), deleteFunction);
    grad_fn->set_next_edges(collect_next_edges( self, other ));
    if (grad_fn->should_compute_output(1)) {
      grad_fn->self_ = SavedVariable(self.clone(), false);
    }
    if (grad_fn->should_compute_output(0)) {
      grad_fn->other_ = SavedVariable(other, false);
    }
  }
  ...
}
```

**Comment by apaszke** [comment]
TensorGeometry seems like a nice idea, I'll try that in a next patch!

**Comment by apaszke** [comment]
Failures are either timeouts, or are data loader multiprocess tests, which I doubt are affected by this change.

**Comment by gchanan** [comment]
I'm guessing this doesn't also magically fix https://github.com/pytorch/pytorch/issues/15115, but I'm going to check anyway.

File: test/test_autograd.py [hunk 1]
```diff
@@ -818,6 +818,21 @@ def scope():
...
+
+        z = x.sum()
+        base_mem = torch.cuda.memory_allocated()
+        z = ((x + 2) * m).sum()
+        end_mem = torch.cuda.memory_allocated()
+
+        # In the end the memory usage should remain equal, because neither of
+        # (x + 2) and ((x + 2) * m) should be kept alive for backward, while the
+        # previous allocation of z had the same size as the current one.
+        self.assertEqual(base_mem, end_mem)
```
Comment (gchanan) [inline test/test_autograd.py]: can you add a test of https://github.com/pytorch/pytorch/issues/15115 -- it's probably more robust than this test because it doesn't rely on memory allocations and actually tests something different (i.e. you could solve https://github.com/pytorch/pytorch/issues/15115 by saving but not unpacking, whereas you can't solve the memory issues that way).

Comment (apaszke) [reply test/test_autograd.py]: Sure, I'll add that test, but I still think this one is robust and tests an important thing. The reason why I took up this patch is because we were using on the order of gigabytes more memory for CNNs that used in-place Dropout2d (which is effectively `input * mask`) which should never happen!

Comment (apaszke) [reply test/test_autograd.py]: Also, note that this patch *doesn't change the unpacking behavior*. So we will still run all the unpacks, except that variables that weren't saved will unpack as undefined tensors (which is fine, because they won't be used anyway).

Comment (gchanan) [reply test/test_autograd.py]: right, I'm not saying one test or the other is better, I'm saying they test different things and should both be tested.

**Review by gchanan — APPROVED** [review]

**Comment by gchanan** [comment]
I don't know why there are failing tests.

**Comment by apaszke** [comment]
I don't see any tests failing. Where are those? Can you rerun them? It used to be green.

**Comment by gchanan** [comment]
I see 4 failing tests:
- ci/circleci: binary_linux_conda_2.7_cpu_build, ci/circleci: binary_linux_conda_3.6_cu90_build  I believe are bogus but should be fixed by https://github.com/pytorch/pytorch/pull/16665.
- pr/py2-clang7-rocmdeb-ubuntu16.04, pr/py2-devtoolset7-rocmrpm-centos7.5 which seem to timeout and which I haven't seen on other PRs.

**Comment by gchanan** [comment]
Trying rebase again.

**Comment by gchanan** [comment]
@apaszke this has consistently timed out on rocm builds, across a few rebases.  Can you take a look?

**Comment by apaszke** [comment]
How can I debug this? It adds a bit of code in `VariableType.cpp`, so if ROCm compiler is *extremely* slow for some reason, we might start overflowing the time if it was close...

**Comment by gchanan** [comment]
@apaszke sorry I should have said RoCM tests (the error message says build, but it's on a test).

Here is the latest example:
```
 18:04:13 test_zeros_like (test_sparse.TestCudaUncoalescedSparse) ... Build timed out (after 120 minutes). Marking the build as failed.
```

**Comment by apaszke** [comment]
Still, do we have instructions that would let me reproduce a ROCm build?
